src/puppetlabs/puppetdb/command.clj

```diff
@@ -428,20 +414,27 @@
         (recur (dec n))
         result))))
 
-(defn process-command-and-respond! [cmd db response-pub-chan stats-atom]
+(defn process-command-and-respond! [{:keys [callback] :as cmd} db response-pub-chan stats-atom]
   (try
     (let [result (call-with-quick-retry 4
                   (fn []
                     (process-command! cmd db)))]
       (swap! stats-atom update :executed-commands inc)
+      (callback {:command cmd :result result})
       (async/>!! response-pub-chan
                  (make-cmd-processed-message cmd nil))
       result)
     (catch Exception ex
+      (callback {:command cmd :exception ex})
       (async/>!! response-pub-chan
                  (make-cmd-processed-message cmd ex))
       (throw ex))))
 
+(defn upgrade-activemq [config enqueue-fn dlo]
+  (when (mig/needs-upgrade? config)
+    (mig/activemq->stockpile config enqueue-fn dlo)
+    (mig/lock-upgrade config)))
```

Comment:
I suppose right now, this means that if some exception that makes it back up to here, we'll retry the upgrade again next time; might be fine (and if that makes it possible for a user to get stuck in a cycle, suppose we can always tell them to touch the file themselves) -- just noting.